Description:  Xext: "xauth generate" with large timeout crashes Xorg #27134
 The security timout CARD32 millis would overflow with large timeouts, causing 
 the timer callback to be called immediately and hit an assertion in
 SecurityAuthorizationExpired.
 .
 This has not been applied upstream yet.  Discussion of the patch faltered
 without resolution.  This patch has a problem when the epoch time is
 sufficiently far in the future (about 2035) that it doesn't matter right now.
Origin: http://patchwork.freedesktop.org/patch/242/
Bug: https://bugs.freedesktop.org/show_bug.cgi?id=27134
Bug-Ubuntu: https://launchpad.net/bugs/519049

diff --git a/Xext/security.c b/Xext/security.c
index af8d205..e81e560 100644
--- a/Xext/security.c
+++ b/Xext/security.c
@@ -279,10 +279,10 @@ SecurityComputeAuthorizationTimeout(
     /* maxSecs is the number of full seconds that can be expressed in
      * 32 bits worth of milliseconds
      */
-    CARD32 maxSecs = (CARD32)(~0) / (CARD32)MILLI_PER_SECOND;
+    CARD32 maxSecs = (CARD32)(MAXINT) / (CARD32)MILLI_PER_SECOND;
 
     if (seconds > maxSecs)
-    { /* only come here if we want to wait more than 49 days */
+    { /* only come here if we want to wait more than 24 days */
 	pAuth->secondsRemaining = seconds - maxSecs;
 	return maxSecs * MILLI_PER_SECOND;
     }
@@ -320,8 +320,6 @@ SecurityAuthorizationExpired(
 {
     SecurityAuthorizationPtr pAuth = (SecurityAuthorizationPtr)pval;
 
-    assert(pAuth->timer == timer);
-
     if (pAuth->secondsRemaining)
     {
 	return SecurityComputeAuthorizationTimeout(pAuth,
